File: frontend/src/api/plan.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "./client";
import type { DishListItem } from "./types";

export interface PlanSlot {
  id: number;
  name: string;
  position: number;
  is_default: boolean;
}

export interface PlanEntry {
  id: number;
  date: string;
  slot_id: number;
  servings: number;
  dish: DishListItem;
}

export interface DayNutrition {
  day: string;
  kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  incomplete: boolean;
}

export interface NutritionTotals {
  kcal: number;
  protein_g: number;
  carbs_g: number;
  fat_g: number;
  incomplete: boolean;
}

export interface Plan {
  start: string;
  days: number;
  entries: PlanEntry[];
  day_sums: DayNutrition[];
  totals: NutritionTotals;
}

export interface Settings {
  daily_kcal_target: number;
  daily_protein_target_g: number;
  daily_carbs_target_g: number;
  daily_fat_target_g: number;
  default_max_missing: number;
}

export function usePlan(start: string, days = 7) {
  return useQuery({
    queryKey: ["plan", start, days],
    queryFn: async () => {
      const { data } = await api.get<Plan>("/plan", { params: { start, days } });
      return data;
    },
  });
}

export function useSettings() {
  return useQuery({
    queryKey: ["settings"],
    queryFn: async () => {
      const { data } = await api.get<Settings>("/settings");
      return data;
    },
  });
}

export function useUpdateSettings() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (payload: Settings) => {
      const { data } = await api.put<Settings>("/settings", payload);
      return data;
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["settings"] }),
  });
}

export function useSlots() {
  return useQuery({
    queryKey: ["plan-slots"],
    queryFn: async () => {
      const { data } = await api.get<PlanSlot[]>("/plan/slots");
      return data;
    },
  });
}

export function useCreateSlot() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (name: string) => {
      const { data } = await api.post<PlanSlot>("/plan/slots", { name });
      return data;
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["plan-slots"] }),
  });
}

export function useMoveSlot() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({ slotId, direction }: { slotId: number; direction: "up" | "down" }) => {
      const { data } = await api.post<PlanSlot[]>(`/plan/slots/${slotId}/move`, { direction });
      return data;
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["plan-slots"] }),
  });
}

export function useDeleteSlot() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (slotId: number) => {
      await api.delete(`/plan/slots/${slotId}`);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ["plan-slots"] });
      queryClient.invalidateQueries({ queryKey: ["plan"] });
    },
  });
}

export function useCreatePlanEntry() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (payload: {
      date: string;
      slot_id: number;
      dish_id: number;
      servings?: number;
    }) => {
      const { data } = await api.post<PlanEntry>("/plan/entries", {
        servings: 1,
        ...payload,
      });
      return data;
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["plan"] }),
  });
}

export function useDeletePlanEntry() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (entryId: number) => {
      await api.delete(`/plan/entries/${entryId}`);
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["plan"] }),
  });
}

export interface GeneratedEntry {
  date: string;
  slot_id: number;
  dish: DishListItem;
}

export function useGeneratePlan() {
  return useMutation({
    mutationFn: async (payload: { start: string; days?: number; slot_ids: number[] }) => {
      const { data } = await api.post<{ entries: GeneratedEntry[] }>("/plan/generate", payload);
      return data.entries;
    },
  });
}

export function useApplyPlan() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async (
      entries: { date: string; slot_id: number; dish_id: number; servings?: number }[],
    ) => {
      const { data } = await api.post<{ applied: number }>("/plan/apply", {
        entries: entries.map((entry) => ({ servings: 1, ...entry })),
      });
      return data;
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["plan"] }),
  });
}

/** Move an entry to another slot by updating it in place. */
export function useMovePlanEntry() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: async ({
      entryId,
      date,
      slot_id,
    }: {
      entryId: number;
      date: string;
      slot_id: number;
    }) => {
      const { data } = await api.patch<PlanEntry>(`/plan/entries/${entryId}`, {
        date,
        slot_id,
      });
      return data;
    },
    onSuccess: () => queryClient.invalidateQueries({ queryKey: ["plan"] }),
  });
}

File: backend/app/schemas/plan.py
from datetime import date as date_type

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.dish import DishListItem


class PlanSlotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    position: int
    is_default: bool


class PlanSlotIn(BaseModel):
    name: str = Field(min_length=1, max_length=50)


class PlanSlotMove(BaseModel):
    direction: str = Field(pattern="^(up|down)$")


class PlanEntryIn(BaseModel):
    date: date_type
    slot_id: int
    dish_id: int
    servings: int = Field(default=1, ge=1, le=50)


class PlanEntryUpdate(BaseModel):
    date: date_type | None = None
    slot_id: int | None = None
    dish_id: int | None = None
    servings: int | None = Field(default=None, ge=1, le=50)


class PlanEntryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    date: date_type
    slot_id: int
    servings: int
    dish: DishListItem


class DayNutritionOut(BaseModel):
    day: date_type
    kcal: float
    protein_g: float
    carbs_g: float
    fat_g: float
    incomplete: bool


class NutritionTotalsOut(BaseModel):
    kcal: float
    protein_g: float
    carbs_g: float
    fat_g: float
    incomplete: bool


class PlanOut(BaseModel):
    start: date_type
    days: int
    entries: list[PlanEntryOut]
    day_sums: list[DayNutritionOut]
    totals: NutritionTotalsOut


class SettingsOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    daily_kcal_target: int
    daily_protein_target_g: int
    daily_carbs_target_g: int
    daily_fat_target_g: int
    default_max_missing: int


class SettingsIn(BaseModel):
    daily_kcal_target: int = Field(ge=500, le=10000)
    daily_protein_target_g: int = Field(ge=0, le=1000)
    daily_carbs_target_g: int = Field(ge=0, le=2000)
    daily_fat_target_g: int = Field(ge=0, le=1000)
    default_max_missing: int = Field(ge=0, le=10)
